tools/block-generator/generator/generate.go

```diff
@@ -875,35 +1120,206 @@ func (g *generator) txnForRound(round uint64) uint64 {
 	return g.config.TxnPerBlock
 }
 
-// finishRound tells the generator it can apply any pending state.
-func (g *generator) finishRound(txnCount uint64) {
-	g.txnCounter += txnCount
+// startRound updates the generator's txnCounter based on the latest block header.
+// It is assumed that g.round has already been incremented in finishRound()
+func (g *generator) startRound() error {
+	if g.round == 0 {
+		// nothing to do in round 0
+		return nil
+	}
+
+	latestHeader, err := g.ledger.BlockHdr(basics.Round(g.round - 1))
+	if err != nil {
+		return fmt.Errorf("Could not obtain block header for round %d: %w", g.round, err)
+	}
+	g.txnCounter = latestHeader.TxnCounter
+	return nil
+}
 
+// finishRound tells the generator it can apply any pending state and updates its round
+func (g *generator) finishRound() {
 	g.timestamp += consensusTimeMilli
 	g.round++
 
 	// Apply pending assets...
 	g.assets = append(g.assets, g.pendingAssets...)
 	g.pendingAssets = nil
 
-	// Apply pending apps...
```

Comment:
A comment with some more explanation about whats going on here would be helpful.

I don't think the loops and cleverness around maps & kinds are worth it here. It hurts clarity and I'm not even sure if it's shorter code.

Did you do it this way because you're imagining lots of additional kinds?

Reply:
The logic around keeping track of the state of each application in the ledger is the most complex part of this PR. It's about to get significantly more complex as being able to generate an AMM swap requires 7 setup steps prior and therefore much more ability to keep track of state. It's worth a discussion.

Reply:
This depends on where the AMM ends up, so I would defer as a future discussion. I've created a TODO for this.